devito/at_controller.py

```diff
@@ -0,0 +1,225 @@
+import math
+import random
+from os import mkdir, path
+
+import cpuinfo
+
+import logger
+
+# global vars
+final_report_name = "final_report.txt"
+default_at_dir = path.join(path.dirname(path.realpath(__file__)), "At Report")
+model_descr_template = "%s %d %d %s %s"
+
+
+class AutoTuner(object):
+
+    def __init__(self, operator, at_report_dir=None):
+        """
+        Object responsible for auto tuning block sizes.
+        :param operator: Operator object.
+        :param at_report_dir: string - indicating path to auto tuning report directory.
+                              If not set default at report directory is used
+        :raises ValueError: if operator is not of Operator type
+        """
+        global final_report_name, default_at_dir
+
+        from operator import Operator
+        if not isinstance(operator, Operator):
+            raise ValueError("Auto tuner requires Operator object to be passed as an argument")
+
+        self.op = operator
+        self.op.propagator.profile = True
+        self.op.apply()  # want to run function at least once to set the arguments
+
+        self.report_dir = at_report_dir if at_report_dir is not None else default_at_dir
+        self.final_report_path = path.join(self.report_dir, final_report_name)
+
+        if not path.isdir(self.report_dir):  # Creates report dir if does not exist
+            mkdir(self.report_dir)
+
+    @staticmethod
+    def get_at_block_size(f_name, time_order, spc_border, shape, block_dims, at_report_dir=None):
+        """
+        Gets block size from auto tuning report
+        :param f_name: str - function name. Used for naming
+        :param time_order: int - time order of kernel
+        :param spc_border: int - space border of kernel
+        :param shape: list - shape of the data buffer
+        :param block_dims: list - indicating which dims are blocked.
+        :param at_report_dir: string - indicating path to auto tuning report directory.
+                              If not set default at report directory is used
+        :raises ValueError: if auto tuning report is not found
+        """
+        global final_report_name, default_at_dir, model_descr_template
+
+        report_dir = at_report_dir if at_report_dir is not None else default_at_dir
+        final_report_path = path.join(report_dir, final_report_name)
```

Comment:
Can the file path not be set as a property in the constructor instead of re-deriving it?